pages/en/lb4/Defining-the-API-using-bottom-up-approach.md

```diff
@@ -0,0 +1,156 @@
+---
+lang: en
+title: 'Defining the API using bottom up approach'
+keywords: LoopBack 4.0, LoopBack 4
+tags:
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Defining-the-API-using-bottom-up-approach.html
+summary:
+---
+
+## Define the API from bottom to top
+
+### Start with LoopBack artfiacts
+
+Sometimes, it's hard to know what your API is going to look like without
+writing out your application first. It also might be difficult to adjust your
+API as plans change or as features need to be reworked.
+
+With TypeScript's [experimental decorator](https://www.typescriptlang.org/docs/handbook/decorators.html)
+feature, APIs can be automatically built and exposed as your application
+continues development. Some key concepts utilize decorators to gather
+'metadata' about your code and then assemble them into a valid OpenAPI spec.
+These concepts and their decorators include:
+
+- [Model](Model.html)
+  - `@model()`
+  - `@property()`
+- [Routes](Routes.html)
+  - `@operation()`
+  - `@param()`
+
+### Define your models
+
+Your models will be used to provide schemas to validate against the data
+your API will intercept and consequently are likely to be referenced as
+argument types in your controllers, so they should be the first to be defined.
+
+{% include note.html content="
+  `Todo` model from [tutorial](TUTORIAL LINK) is used for demonstration here.
+" %}
+
+First, write a simple TypeScript class describing your model and its
+properties.
+
+```ts
+export class Todo {
+  id?: number;
+  title: string;
+  desc?: string;
+  isComplete: boolean;
+}
+```
+
+This model is just a TypeScript class without any decorators in its current
+form, so features such as automatic schema generation are not available to it
+yet.
+
+Now decorate the class with `@model` and `@property`.
+
+```ts
+import {model, property} from '@loopback/repository';
+
+@model()
+export class Todo {
+  @property() id?: number;
+  @property({
+    required: true
+  })
+  title: string;
+  @property() desc?: string;
+  @property() isComplete: boolean;
+}
+```
+
+Once the model has been decorated with these decorators,
+its metadata is made available to the app and the model is ready to have its
+corresponding OpenAPI schema generated.
+
+### Define your routes
+
+{% include note.html content="
+  `TodoController` from [tutorial](TUTORIAL LINK) is used for
+  demonstration here.
+" %}
+
+Once your models are defined, create a controller to host your routes
+for each categories of your API.
+
+```ts
+class TodoController {
+  constructor() {}
+
+  async createTodo(todo: Todo) {
+    // data creating logic goes here
+  }
+
+  async findTodoById(id: number, items?: boolean): Promise<Todo> {
+    // data retrieving logic goes here
+  }
+
+  // ...
+}
+```
+
+The controller's routes in their current state has no information on which
+API endpoints they belong to. Add them in by using `@operation` and `@param`
+decorators.
+
```

Comment:
There should be a tag here to indicate the name of this example file.
`{% include code-caption.html content="src/todo.controller.ts" %}`